### PR TITLE
[ConstraintSystem] Improve type parameter requirement locators

### PR DESCRIPTION
**Pull request by xedin**
When opening generic types with type parameter requirements,
add information about requirement location to the locator of each
generated constraint to make it easier to extract such information
if needed.

### CODE REVIEW

**Comment by xedin** [comment]
@DougGregor does this seem reasonable to you?

File: lib/Sema/ConstraintLocator.h [hunk 1]
```diff
@@ -319,6 +323,11 @@ class ConstraintLocator : public llvm::FoldingSetNode {
       assert(getAssociatedType() == decl);
     }
 
+    PathElement(const swift::Requirement &req, unsigned pos)
```
Comment (DougGregor) [inline lib/Sema/ConstraintLocator.h]: `req` isn't actually used here; can you instead add a `static PathElement getTypeRequirement(unsigned position)`?

Comment (xedin) [reply lib/Sema/ConstraintLocator.h]: Will do, thanks!

**Comment by xedin** [comment]
@swift-ci please smoke test and merge

**Pull request merged**